File: frontend/src/types/player.ts
export interface PlayerMetrics {
  offensive_metric: number | null
  defensive_metric: number | null
  overall_metric: number | null
  offensive_percentile: number | null
  defensive_percentile: number | null
  composite_score?: number | null
  composite_rank?: number | null
  category_scores?: Record<string, number> | null
}

export interface PlayerTrackingStats {
  touches: number | null
  points_per_touch: number | null
  time_of_possession: number | null
  deflections: number | null
  contested_shots: number | null
}

export interface Player {
  id: number
  nba_id: number
  name: string
  position: string | null
  team_abbreviation: string | null
  metrics: PlayerMetrics | null
}

export interface PlayerPerGameStats {
  id: number
  nba_id: number
  name: string
  position: string | null
  team_abbreviation: string | null
  games_played: number | null
  ppg: number | null
  rpg: number | null
  apg: number | null
  mpg: number | null
  spg: number | null
  bpg: number | null
}

export interface PlayerDetail extends Player {
  season: string
  games_played: number | null
  tracking_stats: PlayerTrackingStats | null
}

export interface GameLog {
  game_date: string | null
  matchup: string | null
  wl: string | null
  minutes: number | null
  points: number | null
  rebounds: number | null
  assists: number | null
  steals: number | null
  blocks: number | null
  turnovers: number | null
  fgm: number | null
  fga: number | null
  fg_pct: number | null
  fg3m: number | null
  fg3a: number | null
  fg3_pct: number | null
  ftm: number | null
  fta: number | null
  ft_pct: number | null
  plus_minus: number | null
  game_score: number | null
  offensive_rebounds: number | null
  defensive_rebounds: number | null
}

export interface PlayerCardOption {
  id: number
  name: string
  position: string | null
  team_abbreviation: string | null
  season: string
}

File: backend/app/schemas/player.py
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class PlayerBase(BaseModel):
    nba_id: int
    name: str
    position: str | None
    team_abbreviation: str | None


class PlayerMetrics(BaseModel):
    offensive_metric: Decimal | None
    defensive_metric: Decimal | None
    overall_metric: Decimal | None
    offensive_percentile: int | None
    defensive_percentile: int | None
    # Present only on the composite /overall leaderboard — weighted z-score
    # aggregate across scoring/playmaking/rebounding/defense/impact.
    composite_score: float | None = None
    composite_rank: int | None = None
    category_scores: dict[str, float] | None = None


class PlayerList(PlayerBase):
    id: int
    metrics: PlayerMetrics | None

    model_config = ConfigDict(from_attributes=True)


class PlayerTrackingStats(BaseModel):
    touches: int | None
    points_per_touch: float | None
    time_of_possession: float | None
    deflections: int | None
    contested_shots: int | None


class PlayerPerGameStats(BaseModel):
    id: int
    nba_id: int
    name: str
    position: str | None
    team_abbreviation: str | None
    games_played: int | None
    ppg: Decimal | None
    rpg: Decimal | None
    apg: Decimal | None
    mpg: Decimal | None
    spg: Decimal | None
    bpg: Decimal | None

    model_config = ConfigDict(from_attributes=True)


class PlayerDetail(PlayerList):
    season: str
    games_played: int | None
    tracking_stats: PlayerTrackingStats | None

    model_config = ConfigDict(from_attributes=True)


class PlayerCardOption(BaseModel):
    """A single player+season entry for the player card selector dropdown."""

    id: int
    name: str
    position: str | None
    team_abbreviation: str | None
    season: str

    model_config = ConfigDict(from_attributes=True)


class PlayerCardOptionPage(BaseModel):
    """Paginated envelope for ``/players/available``.

    Only returned when the caller opts in by passing ``limit`` or ``offset``.
    The bare-array legacy shape is kept for callers that pass neither.

    TODO(api-v2): drop the legacy bare-array shape once the frontend is
    migrated to always pass pagination params. Tracked in follow-up ticket.
    """

    items: list[PlayerCardOption]
    total: int
    limit: int
    offset: int


class PlayerSearchResult(BaseModel):
    """A single fuzzy-search hit on players.name."""

    id: int
    nba_id: int
    name: str
    position: str | None
    team_abbreviation: str | None
    # Only populated when the pg_trgm backend is active; None on the
    # ilike fallback path.
    similarity: float | None = None

    model_config = ConfigDict(from_attributes=True)
